lib/spack/spack/environment.py

```diff
@@ -845,6 +849,32 @@ def concretize(self, force=False):
             self.concretized_order = []
             self.specs_by_hash = {}
 
+        if self.concretize_together:
+            return self._concretize_together()
+
+        return self._concretize_separately()
+
+    def _concretize_together(self):
+        # Exit early if the set of concretized specs is the set of user specs
+        user_specs_did_not_change = not bool(
+            set(self.user_specs) - set(self.concretized_user_specs)
+        )
+        if user_specs_did_not_change:
+            return []
+
+        self.concretized_user_specs = []
+        self.concretized_order = []
+        self.specs_by_hash = {}
```

Comment:
This is an interesting approach and I think is a perfectly valid way to use environments but I think the docs need more updates to account for this alternative behavior. For example from https://spack.readthedocs.io/en/latest/environments.html#environments

> specs are only re-concretized when the user explicitly asks for it

I think that line can remain the same but I think you should highlight that `concretize_together: True` is an exception to this rule in the "Concretizing" section (i.e. state that when this is set, prior specs added by the user and concretized earlier will be reconcretized and potentially different).

Also, `concretize_together` is almost always in conflict with Spec matrix specifications (at https://spack.readthedocs.io/en/latest/environments.html#spec-matrices) so I think it would also be worth stating that in the docs: IMO it would make sense to state that where `concretize_together` is mentioned.

Reply:
> Also, concretize_together is almost always in conflict with Spec matrix specifications

On this account we were discussing with @becker33 a minor modification that I'll also implement here. The idea is to rename the `concretize_together` entry in `spec.yaml` just `concretization` and have it accept an enumeration of strings instead of a bool. 

We can start here with two allowed values e.g. `separate` (default) and `together` and extend it later to more strategies as soon as they are available. This won't put us in a corner with respect to future development of Spec matrices and will let us control the concretization strategy nicely from that single entry in the configuration file..

Reply:
Should be done.